src/kern/npf_alg_pptp.c

```diff
@@ -0,0 +1,763 @@
+			gre_npc.npc_l4.pptp_gre_ctx = &gre_con->ctx;
+			/* setup ip addresses */
+			npf_nat_getorig(nt, &o_addr, &o_port);
+			gre_npc.npc_ips[NPF_SRC] = o_addr;
+			gre_npc.npc_ips[NPF_DST] = npc->npc_ips[NPF_SRC];
+			/* establish gre connection state and associate nat */
+			npfa_pptp_gre_establish_gre_conn(&gre_npc, PFIL_OUT, gre_con, nt);
+		}
+		break;
+
+	case PPTP_CALL_DISCONNECT_NOTIFY:
+		if (pptp->len < sizeof(struct pptp_msg_hdr))
+			return false;
+		npf_nat_getorig(nt, &o_addr, &o_port);
+
+		/* lookup for a gre connection entry */
+		gre_con = npfa_pptp_alg_arg_lookup_with_server_call_id(alg_arg,
+				  pptp->call_id);
+		if (gre_con == NULL)
+			return false;
+
+		npfa_pptp_expire_gre_con(npc->npc_ctx, gre_con, o_addr->word32[0],
+				  npc->npc_ips[NPF_SRC]->word32[0]);
+		break;
+
+	case PPTP_WAN_ERROR_NOTIFY:
+		if (pptp->len < sizeof(struct pptp_msg_hdr))
+			return false;
+
+		NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+				  "pptp wan error notify received, peer's call_id %hu\n",
+				  pptp->call_id);
+
+		gre_con = npfa_pptp_alg_arg_lookup_with_client_call_id(alg_arg,
+				  pptp->call_id);
+		if (gre_con == NULL)
+			return false;
+
+		old_call_id = pptp->call_id;
+		pptp->call_id = gre_con->orig_client_call_id;
+		cksum = tcp->check;
+		tcp->check = npf_fixup16_cksum(cksum, old_call_id,
+				  gre_con->orig_client_call_id);
+		break;
+
+	default:
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ *
+ */
+static void
+npfa_pptp_tcp_destroy(npf_t *npf, npf_conn_t *con)
+{
+	struct pptp_gre_con *gre_con;
+	struct pptp_alg_arg *alg_arg;
+	npf_connkey_t *fw;
+	uint32_t client_ip, server_ip;
+	int i;
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "pptp tcp alg destroy a tcp connection %p\n", con);
+
+	alg_arg = (struct pptp_alg_arg *)npf_nat_get_alg_arg(con->c_nat);
+	fw = npf_conn_getforwkey(con);
+
+	/* only ipv4 is supported */
+	if (NPF_CONNKEY_ALEN(fw) == sizeof(uint32_t) || alg_arg == NULL)
+		return;
+
+	client_ip = fw->ck_key[2];
+	server_ip = fw->ck_key[3];
+
+	NPF_DPRINTFCL(NPF_DC_PPTP_ALG, 50,
+			  "pptp tcp alg destroy a tcp connection %p, "
+			  "client ip %u, server ip %u\n",
+			  con, client_ip, server_ip);
+
+	for (i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		gre_con = &alg_arg->gre_cons[i];
+		if ((gre_con->flags & PPTP_ALG_FL_FREE_ENTRY) == 0)
+			npfa_pptp_expire_gre_con(npf, gre_con, client_ip, server_ip);
+	}
+
+	kmem_free(alg_arg, sizeof(struct pptp_alg_arg));
+}
+
+/*
+ * PPTP GRE ALG translator.
+ */
+static bool
+npfa_pptp_gre_translate(npf_cache_t *npc, npf_nat_t *nt, bool forw)
+{
+	nbuf_t *nbuf = npc->npc_nbuf;
+	struct pptp_gre_hdr *gre;
+	struct pptp_gre_con gre_con;
+
+	if (forw || !npf_iscached(npc, NPC_IP4 | NPC_ALG_PPTP_GRE) ||
+			  npc->npc_proto != IPPROTO_GRE)
+		return false;
+
+	nbuf_reset(nbuf);
+	gre = nbuf_advance(nbuf, npc->npc_hlen, sizeof(struct pptp_gre_hdr));
+	if (gre == NULL)
+		return false;
+
+	gre_con.u64 = (uint64_t)npf_nat_get_alg_arg(nt);
```

Comment:
You need to use `nbuf_ensure_contig()` for the size of the header to ensure that it is in the contiguous buffer. Otherwise, accessing the header is not safe.